### ISSUE TITLE
storage: the remap shard appears to be compacted to 1 past the calculated `as_of`

### ISSUE THREAD

**Issue by guswynn** [MaterializeInc/materialize#15402]
### What version of Materialize are you using?

v0.27.0

### How did you install Materialize?

Materialize Cloud

### What is the issue?

Title + log output

### Relevant log output

```source
thread 'timely:work-0' panicked at 'invalid as_of: as_of(Antichain { elements: [1665615358000] }) < since(Antichain { elements: [1665615358001] })', src/storage/src/source/reclock.rs:449:9
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
   2: mz_storage::source::reclock::ReclockOperator::new::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/source/reclock.rs:449:9
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
   4: mz_storage::source::source_reader_pipeline::remap_operator::{{closure}}::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/source/source_reader_pipeline.rs:812:14
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
   6: <futures_util::future::future::fuse::Fuse<Fut> as core::future::future::Future>::poll
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/future/future/fuse.rs:86:37
   7: <core::pin::Pin<P> as core::future::future::Future>::poll
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/future.rs:124:9
   8: <timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G> as mz_timely_util::operators_async_ext::OperatorBuilderExt<G>>::build_async::{{closure}}::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/timely-util/src/operators_async_ext.rs:165:43
   9: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/dataflow/operators/generic/builder_rc.rs:162:26
  10: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/dataflow/operators/generic/builder_raw.rs:204:9
  11: timely::progress::subgraph::PerOperatorState<T>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:684:30
  12: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:344:26
  13: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:310:17
  14: timely::progress::subgraph::PerOperatorState<T>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:684:30
  15: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:344:26
  16: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:310:17
  17: timely::worker::Wrapper::step::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/worker.rs:754:57
  18: core::option::Option<T>::map
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:929:29
  19: timely::worker::Wrapper::step
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/worker.rs:754:26
  20: timely::worker::Worker<A>::step_or_park
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/worker.rs:396:38
  21: mz_storage::storage_state::Worker<A>::run_client
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/storage_state.rs:134:17
  22: mz_storage::storage_state::Worker<A>::run
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/storage_state.rs:110:33
  23: mz_storage::protocol::server::serve::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/protocol/server.rs:94:9
  24: timely::execute::execute::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/execute.rs:287:22
  25: timely_communication::initialize::initialize_from::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/communication/src/initialize.rs:316:33
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
```

**Comment by aljoscha** [comment]
This has two medium-term fixes that should fix it:
 - https://github.com/MaterializeInc/materialize/pull/15645: fixes the actual bug
 - https://github.com/MaterializeInc/materialize/pull/15656: undoes a previous band-aid fix

**Comment by benesch** [comment]
The long term fix is tracked in #15511, so going to call this one closed.
